matrix_generation/main.cpp

```diff
@@ -0,0 +1,296 @@
+#pragma ONCE
+#include "DynamicMatrix.h"
+#include "IsingHamiltonian.h"
+#include<iostream>
+#include<stdio.h>
+#include"slepcsys.h"
+#include<slepceps.h>
+#include"petscsys.h"
+#include<fstream>
+#include <chrono>
+#include <thread>
+#include "vectorlib.cpp"
+
+
+
+static char help[] = "Distributed Hamiltonian Object-Oriented Generator and Solver , by Andrew Grace.";
+
+void performSolve(double JVal, double BxVal, double BzVal, Mat * JMat, Mat * BxMat, Mat * BzMat, bool verbose){
+    
+    int sizeX, sizeY;
+    Mat Sum;
+
+    MatGetSize(*JMat, &sizeX, &sizeY); 
+    MatCreate(PETSC_COMM_WORLD, &Sum);
+    MatSetSizes(Sum, PETSC_DECIDE, PETSC_DECIDE, sizeX, sizeY );
+    MatSetFromOptions(Sum);
+    MatSetUp(Sum);
+    MatZeroEntries(Sum);
+
+
+    MatAssemblyBegin(Sum, MAT_FINAL_ASSEMBLY);
+    MatAssemblyEnd(Sum, MAT_FINAL_ASSEMBLY);    
+    MatAYPX(Sum, JVal, *JMat, DIFFERENT_NONZERO_PATTERN);
+    MatAYPX(Sum, -1.0 * BxVal, *BxMat, DIFFERENT_NONZERO_PATTERN);
+    MatAYPX(Sum, -1.0 * BzVal, *BzMat, DIFFERENT_NONZERO_PATTERN);
+    if(verbose) PetscPrintf(MPI_COMM_WORLD, "Sum calculated.\nCalculating lowest eigenpair...\n");   
+
+    /* Set up SLEPC */
+    Vec imaginary, real;
+    PetscScalar im, re;
+    EPS solver;
+    int nconv;
+
+    /* Initialize vectors */
+    MatCreateVecs(Sum, NULL, &imaginary);
+    MatCreateVecs(Sum, NULL, &real);    
+    EPSCreate(PETSC_COMM_WORLD, &solver);
+    EPSSetOperators(solver, Sum, NULL);
+    EPSSetProblemType(solver, EPS_NHEP);
```

Comment:
The Hamiltonian is Hermitian
```suggestion
    EPSSetProblemType(solver, EPS_HEP);
```